h2o-algos/src/main/java/hex/kmeans/KMeansSimplexSolver.java

```diff
@@ -753,3 +754,91 @@ public void reduce(FindMinimalWeightTask mrt) {
         }
     }
 }
+
+/**
+ * Map Reduce task to assign cluster index based on calculated flow.
+ * If no cluster assigned - assign cluster with minimal distance.
+ * Return number of points in each cluster and changed input frame based on new cluster assignment.
+ */
+class AssignClusterTask extends MRTask<AssignClusterTask> {
+
+    // IN
+    int _constraintsLength;
+    boolean _hasWeightsColumn;
+
+    int _weightIndex;
+    int _distanceIndexStart;
+    int _flowIndexStart;
+    int _oldAssignmentIndex;
+    int _newAssignmentIndex;
+    int _distanceAssignmentIndex;
+    int _dataStopIndex;
+    
+    // OUT
+    int[] _numberOfPointsInCluster;
+    // changed input chunks
+
+    AssignClusterTask(int constraintsLength, boolean hasWeightsColumn, int numCols){
+        _constraintsLength = constraintsLength;
+        _hasWeightsColumn = hasWeightsColumn;
+        _distanceAssignmentIndex = numCols - 3 - constraintsLength;
+        _oldAssignmentIndex = numCols - 2 - constraintsLength;
+        _newAssignmentIndex = numCols - 1 - constraintsLength;
+        _dataStopIndex = numCols - (_hasWeightsColumn ? 1 : 0) - 3 * _constraintsLength - 3;
+        _weightIndex = _dataStopIndex;
+        _distanceIndexStart = _dataStopIndex + (_hasWeightsColumn ? 1 : 0) + 1;
+        _flowIndexStart = numCols - constraintsLength;
+    }
+    
+    public void assignCluster(Chunk[] cs, int row, int clusterIndex, int[] numberOfPointsInClusterLocal){
+        // old assignment
+        cs[_oldAssignmentIndex].set(row, cs[_newAssignmentIndex].at8(row));
+        // new assignment
+        cs[_newAssignmentIndex].set(row, clusterIndex);
+        // distances
+        cs[_distanceAssignmentIndex].set(row, cs[_dataStopIndex + (_hasWeightsColumn ? 1 : 0)  + clusterIndex].at8(row));
+        numberOfPointsInClusterLocal[clusterIndex]++;
+    }
+
+    @Override
+    public void map(Chunk[] cs) {
+        int[] numberOfPointsInClusterLocal = new int[_constraintsLength];
+        for(int i = 0; i < _constraintsLength; i++){
+            numberOfPointsInClusterLocal[i] = 0;
+        }
+        for (int i = 0; i < cs[0].len(); i++) {
+            if (!_hasWeightsColumn || cs[_weightIndex].at8(i) == 1) {
+                boolean assigned = false;
+                for (int j = 0; j < _constraintsLength; j++) {
+                    if (cs[_flowIndexStart + j].at8(i) == 1) {
+                        assignCluster(cs, i, j, numberOfPointsInClusterLocal);
+                        assigned = true;
+                        break;
+                    }
+                }
+                if(!assigned){
+                    double minDistance = Double.MAX_VALUE;
```

Comment:
Can a floating-point math cause eg. infinities to be present?

Reply:
I think this can't happen. But I changed to code to be 100% safe.